### PR TITLE
SPIKE: rotate EC membership

### PR DESCRIPTION
**Pull request by rabi-siddique**
closes: #XXXX
refs: #XXXX

### CODE REVIEW

File: packages/inter-protocol/test/psm/gov-replace-committee.js [hunk 1]
```diff
@@ -6,15 +6,13 @@
  *   (aka big hammer)
  */
 
-import { Fail } from '@endo/errors';
```
Comment (dckc) [inline packages/inter-protocol/test/psm/gov-replace-committee.js]: `Fail` is available on the `assert` global:

```js
const { Fail } = assert;
```

you might want to add...
```js
/// <reference types="@agoric/vats/src/core/core-eval-env"/>
/// <reference types="@agoric/vats/src/core/types-ambient"/>
```

We should probably document this in https://docs.agoric.com/guides/coreeval/permissions.html#declaring-required-capabilities

File: packages/inter-protocol/test/psm/gov-replace-committee.js [hunk 2]
```diff
@@ -26,15 +24,7 @@ const { values } = Object;
 /** @type {<X, Y>(xs: X[], ys: Y[]) => [X, Y][]} */
 const zip = (xs, ys) => xs.map((x, i) => [x, ys[i]]);
 
-/**
- * @param {ERef<import('@agoric/vats').NameAdmin>} nameAdmin
```
Comment (dckc) [inline packages/inter-protocol/test/psm/gov-replace-committee.js]: Rather than removing all the type declarations, just change the `import(...)` syntax to `@import`. That is:

```js
/**
 * @import { NameAdmin } from '@agoric/vats';
 */
```

and then:

```js
/**
 * @param { NameAdmin } nameAdmin
 */
```

File: packages/inter-protocol/test/psm/gov-replace-committee.js [hunk 3]
```diff
@@ -55,7 +45,7 @@ const reserveThenGetNamePaths = async (nameAdmin, paths) => {
 
   return Promise.all(
     paths.map(async path => {
-      Array.isArray(path) || Fail`path ${path} is not an array`;
+      Array.isArray(path) || 'WWWW';
```
Comment (dckc) [inline packages/inter-protocol/test/psm/gov-replace-committee.js]: As noted above, we can keep `Fail`.

**Review by dckc — COMMENTED** [review]
we can make these changes less invasive...

**Comment by dckc** [comment]
For the title, consider "SPIKE: rotate EC membership"

This exploratory work is great, but I expect we'll want to break it into a number of PRs when we add tests and such.

File: packages/governance/src/committee.js [hunk 4]
```diff
@@ -130,9 +130,9 @@ export const start = (zcf, privateArgs, baggage) => {
     ElectorateCreatorI,
     {
       getPoserInvitation() {
-        return getPoserInvitation(zcf, async (voteCounter, questionSpec) =>
-          creatorFacet.addQuestion(voteCounter, questionSpec),
-        );
+        return getPoserInvitation(zcf, async (voteCounter, questionSpec) => {
+          return creatorFacet.addQuestion(voteCounter, questionSpec);
+        });
```
Comment (Chris-Hibbert) [inline packages/governance/src/committee.js]: I don't consider this an improvement. I sometimes do this conversion in order to be able to add a logging statement within the body, but I try to undo it before checkin. eslint's style rules allow either version.

File: packages/governance/src/committee.js [hunk 5]
```diff
@@ -179,6 +179,9 @@ export const start = (zcf, privateArgs, baggage) => {
       getPublicFacet() {
         return publicFacet;
       },
+      shutdown() {
+        zcf.shutdown('EC Charter Shutdown');
+      },
```
Comment (Chris-Hibbert) [inline packages/governance/src/committee.js]: Let's not add this now. The Committee is the sole holder of the set of questions that have been voted on, and I'd prefer we find a way to write this to vstorage before killing the vats.

The reason to do something like this is to prevent ex-members from voting, so let's do that directly. Try adding a method like `stopAddingVotes()` that sets a flag that blocks use of `addQuestion()` or `getPoserInvitation()`.

File: packages/inter-protocol/test/psm/gov-replace-committee.js [hunk 6]
```diff
@@ -78,44 +88,98 @@ const reserveThenDeposit = async (
...
+      );
+    }
+  }
+};
+
+const inviteECMembers = async (
+  { consume: { namesByAddressAdmin, economicCommitteeCreatorFacet } },
+  { options: { voterAddresses = {}, voterAddressesToRemove = {} } },
 ) => {
+  // Generate Invitations
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/psm/gov-replace-committee.js]: The invitations are generated only once, when the contract starts, so this comment should say "get generated invitations" or something similar.

File: packages/inter-protocol/test/psm/gov-replace-committee.js [hunk 7]
```diff
@@ -78,44 +88,98 @@ const reserveThenDeposit = async (
...
   const distributeInvitations = async addrInvitations => {
     await Promise.all(
       addrInvitations.map(async ([addr, invitationP]) => {
-        const [voterInvitation, charterMemberInvitation] = await Promise.all([
-          invitationP,
-          E(E.get(econCharterKit).creatorFacet).makeCharterMemberInvitation(),
-        ]);
-        console.log('sending charter, voting invitations to', addr);
+        const [voterInvitation] = await Promise.all([invitationP]);
+        trace('sending charter, voting invitations to', addr);
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/psm/gov-replace-committee.js]:
```suggestion
        trace('sending voting invitations to', addr);
```

File: packages/inter-protocol/test/psm/gov-replace-committee.js [hunk 8]
```diff
@@ -178,57 +245,168 @@ const startNewEconomicCommittee = async ({
...
+  );
+  trace('Started new EC Charter Instance Successfully');
+
+  instanceP.reset();
+  instanceP.resolve(E.get(startResult).instance);
+
+  econCharterKit.reset();
+  econCharterKit.resolve(startResult);
+};
+harden(startNewEconCharter);
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/psm/gov-replace-committee.js]: I don't think there's any value in calling `harden` on functions that won't be `export`ed.

File: packages/inter-protocol/test/psm/gov-replace-committee.js [hunk 9]
```diff
@@ -178,57 +245,168 @@ const startNewEconomicCommittee = async ({
...
-  console.log('starting new economic committee:', runConfig);
-  const newElectoratePoser = await startNewEconomicCommittee(permittedPowers);
+const startNewEconCharter = async ({
+  consume: { zoe },
+  produce: { econCharterKit },
+  installation: {
+    consume: { binaryVoteCounter: counterP, econCommitteeCharter: installP },
+  },
+  instance: {
+    produce: { econCommitteeCharter: instanceP },
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/psm/gov-replace-committee.js]: The only reason to rename from `econCommitteeCharter` to something ending in `P` would be if you were going to later handle the promise differently from its resolution. In this case, 
```suggestion
    produce: { econCommitteeCharter },
```

File: packages/inter-protocol/test/psm/gov-replace-committee.js [hunk 10]
```diff
@@ -178,57 +245,168 @@ const startNewEconomicCommittee = async ({
...
+  },
+}) => {
+  const [charterInstall, counterInstall] = await Promise.all([
+    installP,
+    counterP,
+  ]);
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/psm/gov-replace-committee.js]: There's no need to await the charter installation. `startInstance()` takes an `ERef<Installation>`.

File: packages/inter-protocol/test/psm/gov-replace-committee.js [hunk 11]
```diff
@@ -178,57 +245,168 @@ const startNewEconomicCommittee = async ({
...
+
+  const psmKitMap = await psmKit;
+
+  for (const obj of psmKitMap.values()) {
+    await E(creatorFacet).addInstance(
+      obj.psm,
+      obj.psmGovernorCreatorFacet,
+      obj.label,
+    );
+  }
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/psm/gov-replace-committee.js]: Is there any reason not to do this?
```suggestion
  for (const { psm, psmGovernorCreatorFacet, label } of psmKitMap.values()) {
    E(creatorFacet).addInstance(psm, psmGovernorCreatorFacet, label);

    return { publicFacet, creatorFacet };
};
```

File: packages/inter-protocol/test/psm/gov-replace-committee.js [hunk 12]
```diff
@@ -178,57 +245,168 @@ const startNewEconomicCommittee = async ({
...
+    {
+      label: 'auctioneer',
+      instanceP: auctioneer,
+      facetP: E.get(auctioneerKit).governorCreatorFacet,
+    },
+  ];
+
+  for (const { label, instanceP, facetP } of instances) {
+    const instance = await instanceP;
+    const govFacet = await facetP;
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/psm/gov-replace-committee.js]: Even though this is a test, waiting each promise separately can take more time unnecessarily.  Here we can wait for them all at once with `await deeplyFulfilledObject(instances)`, and then we don't need multiple awaits in a loop.

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]

File: packages/governance/src/committee.js (hunk 5 above)
Comment (dckc) [reply packages/governance/src/committee.js]: noted as a design choice in

 - #10137

File: packages/inter-protocol/test/psm/gov-replace-committee.js [hunk 13]
```diff
@@ -178,57 +245,168 @@ const startNewEconomicCommittee = async ({
...
+  }
+};
+
+harden(addGovernorsToEconCharter);
+
+const shutdown = async ({
+  consume: { economicCommitteeCreatorFacet, econCharterKit },
+}) => {
+  console.log('Shutting down old EC Committee');
+  await E(economicCommitteeCreatorFacet).shutdown();
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/psm/gov-replace-committee.js]: This method is available on the new committee, but not the old one, unless the old one gets upgraded.

File: packages/inter-protocol/src/econCommitteeCharter.js [hunk 14]
```diff
@@ -179,6 +180,9 @@ export const start = async (zcf, privateArgs, baggage) => {
       },
       makeCharterMemberInvitation: () =>
         zcf.makeInvitation(charterMemberHandler, INVITATION_MAKERS_DESC),
+      shutdown: () => {
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/src/econCommitteeCharter.js]: If we're going to add this method, it should be tested in unit tests and some test that does an upgrade. Either a bootstrap test or a3p-integration would do.

File: packages/governance/src/committee.js (hunk 5 above)
Comment (Chris-Hibbert) [reply packages/governance/src/committee.js]: If we're going to add this method, it should be tested in unit tests and some test that does an upgrade. Either a bootstrap test or a3p-integration would do.

File: packages/inter-protocol/test/psm/gov-replace-committee.js [hunk 15]
```diff
@@ -178,57 +245,168 @@ const startNewEconomicCommittee = async ({
...
+  console.log('EC Committee shutdown successful');
+
+  console.log('Shutting down old EC Charter');
+  const { creatorFacet } = E.get(econCharterKit);
+  await E(creatorFacet).shutdown();
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/psm/gov-replace-committee.js]: This method is available on the new committee, but not the old one, unless the old one got upgraded.

File: packages/inter-protocol/src/econCommitteeCharter.js [hunk 16]
```diff
@@ -179,6 +179,9 @@ export const start = async (zcf, privateArgs, baggage) => {
...
```
Comment (dckc) [inline packages/inter-protocol/src/econCommitteeCharter.js]: adding shutdown methods to both contracts in 1 commit is awkward - I expect them to land in separate PRs, since the testing considerations are different.

Comment (rabi-siddique) [reply packages/inter-protocol/src/econCommitteeCharter.js]: Good point. I will create two separate PRs along with the tests. 

re: I've created #10151 and #10150 for this purpose.

**Comment by dckc** [comment]
This SPIKE served the purpose of exploring feasibility. Nice work! 

Product development proceeds in other PRs.
